### PR TITLE
Improve saving for inheritance - especially when you have a lot of lan…

### PR DESCRIPTION
**Pull request by ckogler-elements**
Problem: Saving of objects with inherited values takes quite long when you have a lot of data in the objects table .

- only execute the query when childs are available
-  use runtime cache and use ids to do the "slow" "like" ..%" query only once.

### CODE REVIEW

**Comment by dvesh3** [comment]
@ckogler-elements can you please check why tests are failing? thanks!

**Comment by ckogler-elements** [comment]
@dvesh3 :Ähm, no i can't, because i cannot see why the test is failing

"We couldn't display the repository
pimcore/pimcore
This repository may not exist or you may need to use another account to have access to it"

**Comment by weisswurstkanone** [comment]
https://travis-ci.com/github/pimcore/pimcore/jobs/299476777
doesn't work ?

**Comment by weisswurstkanone** [comment]
seems that is broken now for cases where you have a folder or different class inbetween

File: models/DataObject/Concrete/Dao/InheritanceHelper.php [hunk 1]
```diff
@@ -421,43 +421,70 @@ protected function buildTree($currentParentId, $fields = '', $parentIdGroups = n
...
+                 * get the object keys once and reuse it because we don't have a key on o_path and it can take quite long when you have a lot of data
+                 *
+                 */
+                $query = 'SELECT o_id FROM objects WHERE o_path LIKE ' .  $this->db->quote($object->getRealFullPath() . '/%');
 
-            if (!$parentIdGroups) {
-                $result = $this->db->fetchAll($query);
+                if (self::$useRuntimeCache) {
+                    $queryCacheKey = 'tree_ids_'.md5($query);
+                    if(is_null(self::$runtimeCache[$queryCacheKey])){
```
Comment (dvesh3) [inline models/DataObject/Concrete/Dao/InheritanceHelper.php]: check for index
image

**Comment by weisswurstkanone** [comment]
I think hasChildren returns false here as it is taken from the runtime cache

**Comment by weisswurstkanone** [comment]
also see https://github.com/pimcore/pimcore/pull/5550

File: models/DataObject/Concrete/Dao/InheritanceHelper.php [hunk 2]
```diff
@@ -421,43 +421,64 @@ protected function buildTree($currentParentId, $fields = '', $parentIdGroups = n
...
+                }else{
+                    $ids = $this->db->fetchCol($query);
+                }
 
-            if (!$parentIdGroups) {
-                $result = $this->db->fetchAll($query);
 
                 if (isset($params['language'])) {
-                    $result = $this->filterResultByLanguage($result, $params['language'], 'language');
+                    $query = "SELECT a.language as language, b.o_id AS id $fields, b.o_classId AS classId, b.o_parentId AS parentId FROM objects b LEFT JOIN " . $this->storetable . ' a ON b.o_id = a.' . $this->idField . ' WHERE  o_id IN( ' . implode(',',$ids).')'
```
Comment (brusch) [inline models/DataObject/Concrete/Dao/InheritanceHelper.php]: I have a concern regarding the `o_id IN()` condition. If you have hundreds of thousands children, you're probably hitting `max_allowed_packet` .

**Comment by brusch** [comment]
Thanks a lot for your PR! 
Im wondering why there should be a performance difference if both queries are using an index. Can you maybe share some benchmarks (incl. the `EXPLAIN` for each query) of your installation that shows the difference? 
Thanks a lot in advance! 😊

**Comment by ckogler-elements** [comment]
Hi,
sorry - it turned out that it was an Issue with the application. 
The problem was that the fullpath index of the objects table was
UNIQUE INDEX `fullpath` ( `o_key`,`o_path`),
instead of 
UNIQUE INDEX `fullpath` (`o_path`, `o_key`),

so the index wasn't used which caused this slow behaviour. In older versions a index was set on o_path so  this wasn't a problem...

The only thing which could be added is the 
if($object->getDao()->hasChildren([$object::OBJECT_TYPE_OBJECT, $object::OBJECT_TYPE_FOLDER, $object::OBJECT_TYPE_VARIANT],true)){
so that the queries are not executed when no result is possible - but it dosn't realy improve the performace significantly...

**Comment by weisswurstkanone** [comment]
thanks for your investigations especially regarding the other optimization!

**Comment by brusch** [comment]
Alright, thanks a lot for your feedback!

**Pull request closed without merging**